Beacon chain spec

jinfwhuang opened the pull request:
This spec PR proposes two new sub-networks. One is for syncing headers of the beacon state, and the other is for asking questions about the state.

The main part of the PR is about the message formats of each sub-networks. It primarily leverages SSZ. The networking layer follows the portal network protocol.

dapplion commented on beacon-chain/beacon-state-dht.md:
```diff
@@ -0,0 +1,81 @@
+
+
+## TODOs
+- Determine if the message type `BeaconStateProof` needs to be gossiped as well.
```
Each participant of the network will be interested in different parts of the state. Those tracking L1 data would want [`latest_execution_payload_header`](https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/beacon-chain.md#beaconstate) while sharded L2s would want [`shard_buffer`](https://github.com/ethereum/consensus-specs/blob/dev/specs/sharding/beacon-chain.md#beaconstate)

jinfwhuang replied:
Yeah...we can start this message type as a DHT network, and if the request frequency is high and the data recency is important for some use cases for specific part of the state, we could propose a "gossip" network for those data type.

Question, what kind of light client will need constant update of `latest_execution_payload_header`? What purposes do these light clients serve?

dapplion replied:
> Question, what kind of light client will need constant update of latest_execution_payload_header? What purposes do these light clients serve?

Any client wanting to track what we currently call Ethereum mainnet head, after the merge would need to track the beacon chain head and get a `latest_execution_payload_header` proof for every block to track the execution header (current Ethereum mainnet)

jinfwhuang replied:
Yup. That is correct. 

I was wondering what is the end user story about this "light client"? Could this be done on demand instead? 
This might clarifies why I am asking "what kind of clients". https://ethresear.ch/t/beacon-chain-light-client-classification/11061

dapplion commented on beacon-chain/sync-gossip.md:
```diff
@@ -0,0 +1,94 @@
+## Overview
+A beacon chain client could sync committee to perform [state updates](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/sync-protocol.md). The data object [LightClientSnapshot](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/sync-protocol.md#lightclientsnapshot) allows a client to quickly sync to a particular header. Once the client establishes a valid header, it ...
+
+These two data types are placed into separate gossip topics. A light client only needs to connects to `bc-light-client-snapshot` briefly at the start of the sync to get a reasonably recent beacon chain header. The client uses messages in the topic `bc-light-client-update` to advance its header.
```
How would participants query the LightClientSnapshot object? It should include include a merkle proof at a known state root or else you can get a malicious sync committee

dapplion replied:
Gossiping full `LightClientUpdates` every block is quite wasteful since each participant of the network only needs to get and validate the next sync committee once per day.

dapplion replied:
It would be more suitable to request full `LightClientUpdates` on a req resp protocol instead of gossip to reduce bandwidth. Also note that clients need _any_ LightClientUpdate during a specific period, and also this updates are necessary for syncing, so historical updates must be available.

jinfwhuang replied:
> How would participants query the LightClientSnapshot object? It should include include a merkle proof at a known state root or else you can get a malicious sync committee

These are gossip topics. Participants of the gossip channel gets recent updates of `LightClientSnapshot` and `LightClientUpdates`. Participant cannot ask for specific snapshot or update. 

LightClientUpdate contains BLS signatures of the sync committee members. Is this not sufficient? https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/sync-protocol.md#validate_light_client_update

jinfwhuang replied:
> Gossiping full `LightClientUpdates` every block is quite wasteful since each participant of the network only needs to get and validate the next sync committee once per day.

A light client might want to always stay up-to-date. For example, this proposed [light client](https://ethresear.ch/t/a-beacon-chain-light-client-proposal/11064) would want to get constant updates. A light client that only wants to sync once a day can just turn off the p2p participation after its update. I would challenge the usefulness of a light client that only updates once a day. It is always out of sync, what kind of questions can this light client answer? See this post: https://ethresear.ch/t/beacon-chain-light-client-classification/11061

jinfwhuang replied:
> It would be more suitable to request full `LightClientUpdates` on a req resp protocol instead of gossip to reduce bandwidth. Also note that clients need _any_ LightClientUpdate during a specific period, and also this updates are necessary for syncing, so historical updates must be available.

Shouldn't of date light client just sync to the latest head? Light client shouldn't need to sync from genesis or a really old head?

dapplion replied:
The data structure `LightClientUpdates` defined in the spec is not great since it serves two purposes at once. To follow the head on a per block basis you need:
- on every block: sync committee signatures + header
- once per day: next sync committee + proof

If you include the next sync committee in the update `98304 - 1` times that data will be useless to participants

dapplion replied:
There's another complexity to take into account. Individual sync committee members sign the header and gossip their signature. Then, there are aggregators in 4 subnets that aggregate them and gossip to a global topic. Finally the block producer aggregate those aggregates and appends them to a block. The update you are proposing, what aggregate of signatures will it contain? Some bridge node will listen to the individual subnets, the aggregate global topic or the block itself? There can exist multiple `LightClientUpdates` with different permutations of the individual signatures, all valid but resulting in different messages. In the beacon chain libp2p network we guard against this by requiring selection proof to aggregators such that on each slot on average N validators can craft this aggregates

dapplion replied:
> These are gossip topics. Participants of the gossip channel gets recent updates of LightClientSnapshot and LightClientUpdates. Participant cannot ask for specific snapshot or update.

How do you trust the first `LightClientSnapshot` if the light-client has no knowledge of any prior sync committee pubkeys? Will you just blindly store in your state what you get from gossip?

In our Lodestar demo we expect user to get a trusted block root from an explorer, friend or genesis and then craft a LightClientSnapshot with a proof to that block's state root. That's how the client can validate the first LightClientSnapshot and initialize safely

jinfwhuang replied:
I am going to summarize a few things. 

1. How to do initial sync with just information from gossip topics? I think this is possible if snapshot also contains proof. It is not going to be straight forward with regard to security because that could open up possibilities for the the topics to be swarmed with fraudulent snapsshot + updates. 
2. Committee update. As suggested by @dapplion, there should be a message type with proof information associated with the next committee message. This could be DHT network or gossip.
3. How should `LightClientUpdates` work with not fully aggregated sync committee signatures?

From answering these questions, we are noticing a few things that are sub-optimal with how `LightClientSnapshot` and `LightClientUpdate` are designed as message types. 

Let's move this discussion off here. It sounds to me that the scope is quite large. We need to both describe the message types and also the mechanisms of how those messages are processed. Let's collaborate on it.

jinfwhuang replied:
> There's another complexity to take into account. Individual sync committee members sign the header and gossip their signature. Then, there are aggregators in 4 subnets that aggregate them and gossip to a global topic. Finally the block producer aggregate those aggregates and appends them to a block. The update you are proposing, what aggregate of signatures will it contain? Some bridge node will listen to the individual subnets, the aggregate global topic or the block itself? There can exist multiple `LightClientUpdates` with different permutations of the individual signatures, all valid but resulting in different messages. In the beacon chain libp2p network we guard against this by requiring selection proof to aggregators such that on each slot on average N validators can craft this aggregates.
> 
> Note that if you wait for the final data in the block you will track the head constantly 12 seconds late and you can be censored by a single validator

The committee signature contains in the LightClientUpdates has to match the header. LightClientUpdate has the field `head` and `sync_committee_bits`. Those two fields have to be matched. The sync-committee signatures included in the block comes from the global aggregate topic. 

`LightClientUpdate`s come from a bridge node. A bridge node would only construct an update when it has receive a full block, i.e. the block has been proposed.

jinfwhuang commented:
> * add a `beacon-chain/README.md` that explains what the documents in this directory are and probably a disclaimer that makes it clear that this work is currently experimental.

Added

> * _nitpick_: rename `beacon-state-dht.md` to something like `beacon-state-network.md` which is more inline with existing naming conventions of the other spec documents.

Updated.

jinfwhuang commented on beacon-chain/beacon-state-dht.md:
```diff
@@ -0,0 +1,81 @@
+    return serialize(BeaconStateProofKey, proof_key)
+
+
+content_key := get_content_key(root, paths)
+content_id := hash_tree_root(BeaconStateProof, beacon_state_proof)
+payload := serialize(BeaconStateProof, beacon_state_proof)
```
Updated.

jinfwhuang commented on beacon-chain/beacon-state-dht.md:
```diff
@@ -0,0 +1,81 @@
+## DHT Overview
+A client has a trusted beacon state root, and it wants to access some parts of the state. Each of the access request corresponds to some leave nodes of the beacon state. The request is a content lookup on a DHT. The response is a Merkle proof. 
```
I added a short general description. 

I agree that some of the languages are pretty dense right now. I am going to keep them concise for now. The intended audiences are very small for now. Once these specs get more mature, I will add more user friendly high level descriptions.

jinfwhuang commented:
@pipermerriam Do you mind merge this PR? This is a good starting point. I put in clear warning about the the spec's status.

There will be follow-up PRs to improve this.
